### PR TITLE
Http range file v17

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/1576

Describe changes:
- Uses a generic container (a thread-safe hash table whose key is the url/filename) to store file parts with HTTP Range header
- handles reassembly of these parts, while enforcing a global memcap

suricata-verify-pr: 517
https://github.com/OISF/suricata-verify/pull/517

Modifies #6228 with fixing negation
S-V test got improved to cover the code path where the negation was wrong

TODO: 
Renaming to more specific ?
More testing of concurrency ? Is there a generic framework for this ?

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information:
| field | test | baseline | % |
|--|--|--|--|
| | tlpr1_stats_chk |

File: suricata.yaml.in [hunk 1]
```diff
@@ -986,6 +986,12 @@ asn1-max-frames: 256
 #     hashsize: 2048
 
 ##############################################################################
+# Containers default settings
+# containers:
+#   urlrange:
+#     memcap: 100mb
+#     timeout: 60
```
Comment (jasonish) [inline suricata.yaml.in:993]: I don't think there is enough info here to have this in the configuration file as-is.. What is a container from the end-user perspective?

Also, as this is part of HTTP, would it not be better under the applayer/http section?  I know the `container` idea is supposed to be generic, but here's we're specifically configuring file storage for http.

**Comment by jasonish** [comment]
> More testing of concurrency ? Is there a generic framework for this ?

Like some other project?  No that I'm aware of.

**Comment by jasonish** [comment]
> Renaming to more specific ?

With respect to containers, it seems like we're trying to make a generic container, but clearly focused on URL range handling, so the naming is a bit mixed in containers.c which you might think of as being purely generic.  WIth only one user of these containers, I wonder if its too early to try come up with something generic, and maybe it shoulds just be UrlRangeFileContainer or something like that.

**Comment by jasonish** [comment]
I'm able to generate a heap-use-after-free, perhaps due to concurrency.  Here's the ASAN output: https://gist.github.com/jasonish/84c350416407350d21025cdb69100fbe.

I wrote a tool with 3 tests:
- Download a file using ranges sequentially: OK
- Download a file using ranges concurrently: OK
- Download a file using ranges concurrently with duplicated requests (so overlaps): leads to crash above

Here is the test client that triggers the crash: https://github.com/jasonish/suricata-http-range-test/blob/main/src/main.rs#L56

**Comment by jasonish** [comment]
Concurrently or sequential, I was happy to see the last `fileinfo` entry had the details for the complete file:
```json
{"timestamp":"2021-07-20T13:34:24.868491-0600","flow_id":115921792311070,"in_iface":"eno1","event_type":"fileinfo","src_ip":"172.105.5.173","src_port":80,"dest_ip":"10.16.1.11","dest_port":47830,"proto":"TCP","http":{"hostname":"codemonkey.net","url":"/range/10000.iso","http_content_type":"application/octet-stream","content_range":{"raw":"bytes 2000-2999/10000","start":2000,"end":2999,"size":10000},"http_method":"GET","protocol":"HTTP/1.1","status":206,"length":1000},"app_proto":"http","fileinfo":{"filename":"/range/10000.iso","sid":[],"gaps":false,"state":"CLOSED","md5":"ede3d3b685b4e137ba4cb2521329a75e","sha256":"541b3e9daa09b20bf85fa273e5cbd3e80185aa4ec298e765db87742b70138a53","stored":true,"file_id":9,"size":1000,"start":2000,"end":2999,"tx_id":0}}
{"timestamp":"2021-07-20T13:34:24.868872-0600","flow_id":766647992360640,"in_iface":"eno1","event_type":"fileinfo","src_ip":"172.105.5.173","src_port":80,"dest_ip":"10.16.1.11","dest_port":47828,"proto":"TCP","http":{"hostname":"codemonkey.net","url":"/range/10000.iso","http_content_type":"application/octet-stream","content_range":{"raw":"bytes 9000-9999/10000","start":9000,"end":9999,"size":10000},"http_method":"GET","protocol":"HTTP/1.1","status":206,"length":1000},"app_proto":"http","fileinfo":{"filename":"/range/10000.iso","sid":[],"gaps":false,"state":"CLOSED","md5":"ede3d3b685b4e137ba4cb2521329a75e","sha256":"541b3e9daa09b20bf85fa273e5cbd3e80185aa4ec298e765db87742b70138a53","stored":true,"file_id":10,"size":1000,"start":9000,"end":9999,"tx_id":0}}
{"timestamp":"2021-07-20T13:34:24.868872-0600","flow_id":766647992360640,"in_iface":"eno1","event_type":"fileinfo","src_ip":"172.105.5.173","src_port":80,"dest_ip":"10.16.1.11","dest_port":47828,"proto":"TCP","http":{"hostname":"codemonkey.net","url":"/range/10000.iso","http_content_type":"application/octet-stream","content_range":{"raw":"bytes 9000-9999/10000","start":9000,"end":9999,"size":10000},"http_method":"GET","protocol":"HTTP/1.1","status":206,"length":1000},"app_proto":"http","fileinfo":{"filename":"/range/10000.iso","sid":[],"gaps":false,"state":"CLOSED","md5":"a00fa935afb8ed25e9736f64880d6bcc","sha256":"1cc8d4b2763a5dd441968ed2685cf3abe8c4051cacd0422e428c07a720bd4475","stored":true,"file_id":11,"size":10000,"tx_id":0}}
```

I wonder if we need to call more attention to whats the final file?

Also up for discussion with respect to file extraction is should we keep the partials?

File: suricata.yaml.in (hunk 1 above)
Comment (catenacyber) [reply suricata.yaml.in:993]: Indeed, I do not feel very good about the naming. And I hesitated between making something generic or specific...

I will go for making names more specific...

**Comment by catenacyber** [comment]
> I wonder if we need to call more attention to whats the final file?

We can see if a file is partial is final or partial with the `start` field in the json...

> Also up for discussion with respect to file extraction is should we keep the partials?

As for now, we do not reassemble the whole files and only have the partial ones.
So, continuity would be to have both now...

**Comment by catenacyber** [comment]
> Here is the test client that triggers the crash: https://github.com/jasonish/suricata-http-range-test/blob/main/src/main.rs#L56

Thanks so much for this.
I think I understand my bug.
But I cannot reproduce it because I get 301 `"redirect": "https://codemonkey.net/range/10000.iso"`
Am I missing something ?

**Comment by catenacyber** [comment]
Replaced by https://github.com/OISF/suricata/pull/6273

**Pull request closed without merging**